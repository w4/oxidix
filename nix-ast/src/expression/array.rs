use std::iter::Peekable;

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{Expression, HandleStreamError, SpannedError, parse_primary};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct ArrayExpression<'a> {
    pub values: Vec<Expression<'a>>,
}

impl<'a> ArrayExpression<'a> {
    pub fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<ArrayExpression<'a>, SpannedError> {
        let mut values = Vec::new();

        loop {
            let (token, _) = stream.peek().span_error()?;

            if TokenDiscriminants::from(token) == TokenDiscriminants::SquareBracketClose {
                break;
            }

            values.push(parse_primary(stream)?);
        }

        Ok(ArrayExpression { values })
    }
}
