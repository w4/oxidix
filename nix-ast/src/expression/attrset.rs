use std::{collections::BTreeMap, iter::Peekable};

use nix_lexer::{Lexer, Token, TokenDiscriminants};

use crate::{
    BindingName, Error, Expression, expect_next_token_or_error, parse_binding_name,
    parse_expression_inner,
};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct AttrsetExpression<'a> {
    pub recursive: bool,
    pub bindings: BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
}

impl<'a> AttrsetExpression<'a> {
    pub fn parse(stream: &mut Peekable<Lexer<'a, Token<'a>>>) -> Result<Self, Error> {
        let mut bindings = BTreeMap::new();

        loop {
            match stream.peek() {
                Some(Ok(Token::BraceClose)) | None => break,
                Some(Ok(Token::Inherit)) => parse_inherits(stream, &mut bindings)?,
                Some(Ok(Token::Ident(_) | Token::String(_) | Token::InterpolationStart)) => {
                    parse_key_value(stream, &mut bindings)?
                }
                Some(Ok(v)) => {
                    return Err(Error::UnexpectedToken(
                        v.into(),
                        vec![TokenDiscriminants::BraceClose, TokenDiscriminants::Ident],
                    ));
                }
                Some(Err(())) => panic!(),
            }
        }

        expect_next_token_or_error(stream, TokenDiscriminants::BraceClose)?;

        Ok(AttrsetExpression {
            recursive: false,
            bindings,
        })
    }
}

pub fn parse_key_value<'a>(
    stream: &mut Peekable<Lexer<'a, Token<'a>>>,
    bindings: &mut BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
) -> Result<(), Error> {
    let key = parse_binding_name(stream)?;
    let value = parse_expression_inner(stream, u8::MAX)?;
    expect_next_token_or_error(stream, TokenDiscriminants::Semicolon)?;

    if bindings.insert(key.clone(), value).is_some() {
        return Err(Error::AttributeAlreadyDefined(format!("{key:?}")));
    }

    Ok(())
}

pub fn parse_inherits<'a>(
    stream: &mut Peekable<Lexer<'a, Token<'a>>>,
    bindings: &mut BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
) -> Result<(), Error> {
    expect_next_token_or_error(stream, TokenDiscriminants::Inherit)?;

    match stream.peek() {
        Some(Ok(Token::Ident(_))) => parse_inherits_no_from(stream, bindings)?,
        Some(Ok(Token::BracketOpen)) => parse_inherits_with_from(stream, bindings)?,
        Some(Ok(Token::Semicolon)) => {}
        Some(Ok(v)) => {
            return Err(Error::UnexpectedToken(
                v.into(),
                vec![TokenDiscriminants::Ident, TokenDiscriminants::BracketOpen],
            ));
        }
        Some(Err(())) => panic!(),
        None => return Err(Error::UnexpectedEndOfFile),
    };

    Ok(())
}

pub fn parse_inherits_no_from<'a>(
    stream: &mut Peekable<Lexer<'a, Token<'a>>>,
    bindings: &mut BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
) -> Result<(), Error> {
    loop {
        match stream.next() {
            Some(Ok(Token::Ident(v))) => {
                if bindings
                    .insert(vec![BindingName::Value(v)], Expression::Ident(v))
                    .is_some()
                {
                    return Err(Error::AttributeAlreadyDefined(v.to_string()));
                }
            }
            Some(Ok(Token::Semicolon)) => break,
            Some(Ok(v)) => {
                return Err(Error::UnexpectedToken(
                    v.into(),
                    vec![TokenDiscriminants::Ident, TokenDiscriminants::Semicolon],
                ));
            }
            Some(Err(())) => panic!(),
            None => return Err(Error::UnexpectedEndOfFile),
        }
    }

    Ok(())
}

pub fn parse_inherits_with_from<'a>(
    stream: &mut Peekable<Lexer<'a, Token<'a>>>,
    bindings: &mut BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
) -> Result<(), Error> {
    expect_next_token_or_error(stream, TokenDiscriminants::BracketOpen)?;

    let binding = parse_expression_inner(stream, u8::MAX)?;

    expect_next_token_or_error(stream, TokenDiscriminants::BracketClose)?;

    loop {
        match stream.next() {
            Some(Ok(Token::Ident(v))) => {
                if bindings
                    .insert(vec![BindingName::Value(v)], binding.clone())
                    .is_some()
                {
                    return Err(Error::AttributeAlreadyDefined(v.to_string()));
                }
            }
            Some(Ok(Token::Semicolon)) => break,
            Some(Ok(v)) => {
                return Err(Error::UnexpectedToken(
                    v.into(),
                    vec![TokenDiscriminants::Ident, TokenDiscriminants::Semicolon],
                ));
            }
            Some(Err(())) => panic!(),
            None => return Err(Error::UnexpectedEndOfFile),
        }
    }

    Ok(())
}
