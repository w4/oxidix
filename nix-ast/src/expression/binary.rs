//! Binary expression parsing and representation.
//!
//! This module handles all binary operations in Nix, including arithmetic,
//! logical, comparison, and Nix-specific operations like attribute selection
//! and function application. It implements proper operator precedence and
//! associativity according to the Nix language specification.

use std::iter::Peekable;

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{
    Error, Expression, HandleStreamError, SpannedError, expect_next_token_or_error,
    parse_expression_inner,
};

/// Represents a binary operation between two expressions.
///
/// Binary expressions form the backbone of most Nix computations, from simple
/// arithmetic to complex attribute manipulations. This struct handles all
/// binary operations with proper precedence and associativity.
///
/// # Examples
///
/// ```rust
/// // Arithmetic: 1 + 2
/// // Comparison: x == y
/// // Attribute selection: obj.attr
/// // Function application: func arg
/// ```
#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct BinaryExpression<'a> {
    /// The binary operator being applied
    pub operator: BinaryOperator,
    /// The left operand expression
    pub left: Expression<'a>,
    /// The right operand expression
    pub right: Expression<'a>,
    /// Optional postfix expression (used for `or` in attribute selection)
    pub postfix: Option<Expression<'a>>,
}

impl<'a> BinaryExpression<'a> {
    pub fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
        min_prec: u8,
    ) -> Result<Option<(BinaryOperator, Expression<'a>, Option<Expression<'a>>)>, SpannedError>
    {
        let (token, _) = match stream.peek() {
            Some(token) => token.span_error()?,
            None => return Ok(None),
        };

        let Some(operator) = BinaryOperator::parse(token) else {
            return Ok(None);
        };

        let prec = operator.precedence();

        if prec > min_prec {
            return Ok(None);
        }

        // don't chomp the token for juxtaposition-based function application
        if operator != BinaryOperator::FunctionApplication {
            stream.next();
        }

        let next_prec = match operator.associativity() {
            Some(BinaryAssociativity::Left) => prec.saturating_sub(1),
            Some(BinaryAssociativity::Right) => prec,
            None => u8::MAX,
        };

        let right = match operator {
            BinaryOperator::AttributeSelection => parse_binary_expression(stream, next_prec)?,
            _ => parse_expression_inner(stream, next_prec)?,
        };

        let postfix = match operator {
            BinaryOperator::AttributeSelection => {
                if let Some((Ok(Token::TextOr), _)) = stream.peek() {
                    stream.next();
                    Some(parse_expression_inner(stream, next_prec)?)
                } else {
                    None
                }
            }
            _ => None,
        };

        Ok(Some((operator, right, postfix)))
    }
}

fn parse_binary_expression<'a>(
    stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    next_prec: u8,
) -> Result<Expression<'a>, SpannedError> {
    match stream.next().span_error()? {
        (Token::Ident(v), _) => Ok(Expression::Ident(v)),
        (Token::InterpolationStart, _) => {
            let expr = parse_expression_inner(stream, next_prec)?;
            expect_next_token_or_error(stream, TokenDiscriminants::BraceClose)?;
            Ok(expr)
        }
        (Token::String(_), _) => todo!(),
        (v, span) => Err(Error::UnexpectedToken(
            v.into(),
            vec![
                TokenDiscriminants::Ident,
                TokenDiscriminants::InterpolationStart,
                TokenDiscriminants::String,
                TokenDiscriminants::Or,
            ],
        )
        .with_span(span)),
    }
}

/// Binary operators in the Nix language.
///
/// This enum represents all binary operators supported by Nix, ordered by
/// their precedence (lowest to highest). The precedence follows the official
/// Nix language specification.
///
/// # Precedence Order (lowest to highest)
///
/// 1. Attribute selection (`.`)
/// 2. Function application (juxtaposition)
/// 3. Has attribute (`?`)
/// 4. List concatenation (`++`)
/// 5. Multiplication (`*`), Division (`/`)
/// 6. Addition (`+`), Subtraction (`-`)
/// 7. Update (`//`)
/// 8. Comparisons (`<`, `<=`, `>`, `>=`, `==`, `!=`)
/// 9. Logical AND (`&&`)
/// 10. Logical OR (`||`)
/// 11. Logical implication (`->`)
/// 12. Pipe operators (`|>`, `<|`)
///
/// # Reference
///
/// See: https://nix.dev/manual/nix/2.29/language/operators.html
#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord, Copy)]
pub enum BinaryOperator {
    /// Attribute selection operator (`.`) - `obj.attr`
    AttributeSelection,

    /// Function application (juxtaposition) - `func arg`
    FunctionApplication,

    /// Has attribute operator (`?`) - `obj ? attr`
    HasAttribute,

    /// List concatenation operator (`++`) - `list1 ++ list2`
    ListConcatenation,

    /// Multiplication operator (`*`) - `a * b`
    Multiplication,

    /// Division operator (`/`) - `a / b`
    Division,

    /// Subtraction operator (`-`) - `a - b`
    Subtraction,

    /// Addition operator (`+`) - `a + b`
    Addition,

    /// Update operator (`//`) - `set1 // set2`
    Update,

    /// Less than operator (`<`) - `a < b`
    Lt,

    /// Less than or equal operator (`<=`) - `a <= b`
    Le,

    /// Greater than operator (`>`) - `a > b`
    Gt,

    /// Greater than or equal operator (`>=`) - `a >= b`
    Ge,

    /// Equality operator (`==`) - `a == b`
    Eq,

    /// Inequality operator (`!=`) - `a != b`
    Ne,

    /// Logical AND operator (`&&`) - `a && b`
    LogicalAnd,

    /// Logical OR operator (`||`) - `a || b`
    LogicalOr,

    /// Logical implication operator (`->`) - `a -> b`
    LogicalImplication,

    /// Left-associative pipe operator (`|>`) - `a |> b`
    LeftAssocPipeOperator,

    /// Right-associative pipe operator (`<|`) - `a <| b`
    RightAssocPipeOperator,
}

impl BinaryOperator {
    /// Returns the precedence level of this operator.
    ///
    /// Lower numbers indicate higher precedence (tighter binding).
    /// This follows the Nix language specification for operator precedence.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use nix_ast::expression::BinaryOperator;
    ///
    /// assert!(BinaryOperator::Multiplication.precedence() <
    ///         BinaryOperator::Addition.precedence());
    /// ```
    pub fn precedence(self) -> u8 {
        match self {
            Self::AttributeSelection => 1,
            Self::FunctionApplication => 2,
            // 3 = unary arithmetic negation
            Self::HasAttribute => 4,
            Self::ListConcatenation => 5,
            Self::Multiplication | Self::Division => 6,
            Self::Subtraction | Self::Addition => 7,
            // 8 = unary logical NOT
            Self::Update => 9,
            Self::Lt | Self::Le | Self::Gt | Self::Ge | Self::Eq | Self::Ne => 10,
            Self::LogicalAnd => 12,
            Self::LogicalOr => 13,
            Self::LogicalImplication => 14,
            Self::LeftAssocPipeOperator | Self::RightAssocPipeOperator => 15,
        }
    }

    /// Attempts to parse a binary operator from a token.
    ///
    /// This method examines a token and determines if it represents a binary
    /// operator. Some tokens (like keywords and delimiters) cannot be binary
    /// operators, while others default to function application.
    ///
    /// # Arguments
    ///
    /// * `token` - The token to examine
    ///
    /// # Returns
    ///
    /// Returns `Some(BinaryOperator)` if the token represents a binary operator,
    /// or `None` if it cannot be used as a binary operator.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use nix_ast::expression::BinaryOperator;
    /// use nix_lexer::Token;
    ///
    /// assert_eq!(BinaryOperator::parse(&Token::Plus), Some(BinaryOperator::Addition));
    /// assert_eq!(BinaryOperator::parse(&Token::Let), None);
    /// ```
    pub fn parse(token: &Token<'_>) -> Option<Self> {
        match token {
            Token::Dot => Some(Self::AttributeSelection),
            Token::Question => Some(Self::HasAttribute),
            Token::PlusPlus => Some(Self::ListConcatenation),
            Token::Asterisk => Some(Self::Multiplication),
            Token::Slash => Some(Self::Division),
            Token::Neg => Some(Self::Subtraction),
            Token::Plus => Some(Self::Addition),
            Token::SlashSlash => Some(Self::Update),
            Token::Lt => Some(Self::Lt),
            Token::Le => Some(Self::Le),
            Token::Gt => Some(Self::Gt),
            Token::Ge => Some(Self::Ge),
            Token::DoubleEquals => Some(Self::Eq),
            Token::BangEquals => Some(Self::Ne),
            Token::And => Some(Self::LogicalAnd),
            Token::Or => Some(Self::LogicalOr),
            Token::Arrow => Some(Self::LogicalImplication),
            Token::PipeRight => Some(Self::LeftAssocPipeOperator),
            Token::PipeLeft => Some(Self::RightAssocPipeOperator),
            Token::If
            | Token::Then
            | Token::Else
            | Token::Let
            | Token::In
            | Token::Semicolon
            | Token::BraceClose
            | Token::BracketClose
            | Token::SquareBracketClose
            | Token::Comma => None,
            _ => Some(Self::FunctionApplication),
        }
    }

    /// Returns the associativity of this operator.
    ///
    /// Associativity determines how operators of the same precedence are
    /// grouped. Left-associative operators group from left to right,
    /// right-associative operators group from right to left, and some
    /// operators (like comparisons) are non-associative.
    ///
    /// # Returns
    ///
    /// - `Some(BinaryAssociativity::Left)` for left-associative operators
    /// - `Some(BinaryAssociativity::Right)` for right-associative operators  
    /// - `None` for non-associative operators
    ///
    /// # Examples
    ///
    /// ```rust
    /// use nix_ast::expression::{BinaryOperator, BinaryAssociativity};
    ///
    /// // Addition is left-associative: a + b + c = (a + b) + c
    /// assert_eq!(BinaryOperator::Addition.associativity(),
    ///           Some(BinaryAssociativity::Left));
    ///
    /// // Update is right-associative: a // b // c = a // (b // c)
    /// assert_eq!(BinaryOperator::Update.associativity(),
    ///           Some(BinaryAssociativity::Right));
    ///
    /// // Equality is non-associative: a == b == c is not allowed
    /// assert_eq!(BinaryOperator::Eq.associativity(), None);
    /// ```
    pub fn associativity(self) -> Option<BinaryAssociativity> {
        match self {
            Self::AttributeSelection
            | Self::HasAttribute
            | Self::Lt
            | Self::Le
            | Self::Gt
            | Self::Ge
            | Self::Eq
            | Self::Ne => None,
            Self::FunctionApplication
            | Self::Multiplication
            | Self::Division
            | Self::Subtraction
            | Self::Addition
            | Self::LogicalAnd
            | Self::LogicalOr
            | Self::LeftAssocPipeOperator => Some(BinaryAssociativity::Left),
            Self::ListConcatenation
            | Self::Update
            | Self::LogicalImplication
            | Self::RightAssocPipeOperator => Some(BinaryAssociativity::Right),
        }
    }
}

/// Associativity of binary operators.
///
/// This enum determines how operators of the same precedence level are
/// grouped when they appear in sequence. For example, with left associativity,
/// `a + b + c` is parsed as `(a + b) + c`, while with right associativity,
/// `a // b // c` is parsed as `a // (b // c)`.
///
/// Some operators (like comparison operators) are non-associative, meaning
/// expressions like `a == b == c` are not allowed and will result in a
/// parse error.
#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord, Copy)]
pub enum BinaryAssociativity {
    /// Left-to-right associativity: `a op b op c` = `(a op b) op c`
    Left,
    /// Right-to-left associativity: `a op b op c` = `a op (b op c)`
    Right,
}
