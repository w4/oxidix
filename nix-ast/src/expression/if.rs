use std::iter::Peekable;

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{Expression, SpannedError, expect_next_token_or_error, parse_expression_inner};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct IfExpression<'a> {
    pub condition: Expression<'a>,
    pub then: Expression<'a>,
    pub r#else: Expression<'a>,
}

impl<'a> IfExpression<'a> {
    pub fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<IfExpression<'a>, SpannedError> {
        let condition = parse_expression_inner(stream, u8::MAX)?;
        expect_next_token_or_error(stream, TokenDiscriminants::Then)?;
        let then = parse_expression_inner(stream, u8::MAX)?;
        expect_next_token_or_error(stream, TokenDiscriminants::Else)?;
        let r#else = parse_expression_inner(stream, u8::MAX)?;

        Ok(IfExpression {
            condition,
            then,
            r#else,
        })
    }
}
