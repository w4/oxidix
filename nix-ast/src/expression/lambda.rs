use std::{collections::BTreeMap, iter::Peekable};

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{
    Error, Expression, HandleStreamError, SpannedError, expect_next_token_or_error,
    parse_expression_inner,
};

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct LambdaExpression<'a> {
    pub parameters: LambdaParameters<'a>,
    pub body: Expression<'a>,
}

impl<'a> LambdaExpression<'a> {
    pub fn parse(
        orig_stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<Option<LambdaExpression<'a>>, SpannedError> {
        // clone the stream so we can optimistically parse
        let mut stream = orig_stream.clone();

        let Some(parameters) = LambdaParameters::parse(&mut stream).transpose() else {
            return Ok(None);
        };

        let Some((Ok(Token::Colon), _)) = stream.next() else {
            return Ok(None);
        };

        // if we've confirmed we're in a lambda we can return any syntax errors from parsing parameters now
        let parameters = parameters?;

        let body = parse_expression_inner(&mut stream, u8::MAX)?;

        // since we've passed parsing, fast forward the stream that was passed to us
        *orig_stream = stream;

        Ok(Some(LambdaExpression { parameters, body }))
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub enum LambdaParameters<'a> {
    Set(LambdaParametersSet<'a>),
    Named(&'a str),
}

impl<'a> LambdaParameters<'a> {
    fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<Option<Self>, SpannedError> {
        match stream.next() {
            Some((Ok(Token::BraceOpen), _)) => {
                Ok(LambdaParametersSet::parse(stream)?.map(LambdaParameters::Set))
            }
            Some((Ok(Token::Ident(ident)), _)) => Ok(Some(LambdaParameters::Named(ident))),
            _ => Ok(None),
        }
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct LambdaParametersSet<'a> {
    pub inner: BTreeMap<&'a str, LambdaParameter<'a>>,
    pub extra_ignored: bool,
    pub bound: Option<&'a str>,
}

impl<'a> LambdaParametersSet<'a> {
    fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<Option<Self>, SpannedError> {
        let mut inner = BTreeMap::new();
        let mut extra_ignored = false;

        loop {
            let (ident, span) = match stream.next().span_error()? {
                (Token::Ident(ident), span) => (ident, span),
                (Token::BraceClose, _) => break,
                (Token::DotDotDot, _) => {
                    extra_ignored = true;
                    expect_next_token_or_error(stream, TokenDiscriminants::BraceClose)?;
                    break;
                }
                (v, span) => {
                    return Err(Error::UnexpectedToken(
                        v.into(),
                        vec![
                            TokenDiscriminants::Ident,
                            TokenDiscriminants::BraceClose,
                            TokenDiscriminants::DotDotDot,
                        ],
                    )
                    .with_span(span));
                }
            };

            let default = if matches!(stream.peek(), Some((Ok(Token::Question), _))) {
                let _token = stream.next();
                let default = parse_expression_inner(stream, u8::MAX)?;
                Some(default)
            } else {
                None
            };

            if inner.insert(ident, LambdaParameter { default }).is_some() {
                return Err(Error::DuplicateFunctionArgument(ident.to_string()).with_span(span));
            }

            match stream.next().span_error()? {
                (Token::BraceClose, _) => break,
                (Token::Comma, _) => {}
                (v, span) => {
                    return Err(Error::UnexpectedToken(
                        v.into(),
                        vec![
                            TokenDiscriminants::Question,
                            TokenDiscriminants::Comma,
                            TokenDiscriminants::BraceClose,
                        ],
                    )
                    .with_span(span));
                }
            }
        }

        let bound = if matches!(stream.peek(), Some((Ok(Token::At), _))) {
            stream.next();

            match stream.next().span_error()? {
                (Token::Ident(ident), _) => Some(ident),
                (v, span) => {
                    return Err(
                        Error::UnexpectedToken(v.into(), vec![TokenDiscriminants::Ident])
                            .with_span(span),
                    );
                }
            }
        } else {
            None
        };

        Ok(Some(LambdaParametersSet {
            inner,
            extra_ignored,
            bound,
        }))
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct LambdaParameter<'a> {
    pub default: Option<Expression<'a>>,
}
