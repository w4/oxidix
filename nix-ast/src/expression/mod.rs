mod array;
mod attrset;
mod binary;
mod r#if;
mod lambda;
mod r#let;
mod unary;

pub use array::ArrayExpression;
pub use attrset::AttrsetExpression;
pub use binary::BinaryExpression;
pub use r#if::IfExpression;
pub use lambda::LambdaExpression;
pub use r#let::LetExpression;
pub use unary::UnaryExpression;
