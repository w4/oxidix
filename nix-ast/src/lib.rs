mod expression;

use std::{
    iter::Peekable,
    num::{ParseFloatError, ParseIntError},
};

use expression::{
    ArrayExpression, AttrsetExpression, BinaryExpression, IfExpression, LambdaExpression,
    LetExpression, UnaryExpression,
};
use nix_lexer::{Lexer, Token, TokenDiscriminants};
use thiserror::Error;

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub enum Expression<'a> {
    Bool(bool),
    Int(i64),
    Float(WrappedFloat),
    Add(Box<(Expression<'a>, Expression<'a>)>),
    Let(Box<LetExpression<'a>>),
    Ident(&'a str),
    If(Box<IfExpression<'a>>),
    Binary(Box<BinaryExpression<'a>>),
    Unary(Box<UnaryExpression<'a>>),
    Array(ArrayExpression<'a>),
    Comment,
    Lambda(Box<LambdaExpression<'a>>),
    Attrset(AttrsetExpression<'a>),
}

#[derive(Error, Debug)]
pub enum Error {
    #[error("Unexpected end of file")]
    UnexpectedEndOfFile,
    #[error("Unexpected token {0:?}")]
    UnexpectedTopLevelToken(TokenDiscriminants),
    #[error("Unexpected token {0:?} expected one of {1:?}")]
    UnexpectedToken(TokenDiscriminants, Vec<TokenDiscriminants>),
    #[error("Invalid integer: {0}")]
    ParseInt(#[from] ParseIntError),
    #[error("Invalid float: {0}")]
    ParseFloat(#[from] ParseFloatError),
    #[error("Duplicate formal function argument: {0}")]
    DuplicateFunctionArgument(String),
    #[error("Attribute `{0}` already defined")]
    AttributeAlreadyDefined(String),
}

pub fn parse_expression<'a>(stream: Lexer<'a, Token<'a>>) -> Result<Expression<'a>, Error> {
    parse_expression_inner(&mut stream.peekable(), u8::MAX)
}

fn parse_expression_inner<'a>(
    stream: &mut Peekable<Lexer<'a, Token<'a>>>,
    min_prec: u8,
) -> Result<Expression<'a>, Error> {
    let mut left = if let Some(expr) = UnaryExpression::parse(stream)? {
        Expression::Unary(Box::new(expr))
    } else if let Some(lambda) = LambdaExpression::parse(stream)? {
        Expression::Lambda(Box::new(lambda))
    } else {
        parse_primary(stream)?
    };

    loop {
        let Some((operator, right, postfix)) = BinaryExpression::parse(stream, min_prec)? else {
            break;
        };

        left = Expression::Binary(Box::new(BinaryExpression {
            operator,
            left,
            right,
            postfix,
        }));
    }

    Ok(left)
}

fn parse_primary<'a>(stream: &mut Peekable<Lexer<'a, Token<'a>>>) -> Result<Expression<'a>, Error> {
    let token = stream.next().ok_or(Error::UnexpectedEndOfFile)?.unwrap();

    match token {
        Token::Bool(v) => Ok(Expression::Bool(v)),
        Token::Let => Ok(Expression::Let(Box::new(LetExpression::parse(stream)?))),
        Token::If => Ok(Expression::If(Box::new(IfExpression::parse(stream)?))),
        Token::Int(v) => Ok(Expression::Int(v.parse()?)),
        Token::Float(v) => Ok(Expression::Float(WrappedFloat(v.parse()?))),
        Token::Ident(v) => Ok(Expression::Ident(v)),
        Token::BracketOpen => wrapped_interpolated(stream, TokenDiscriminants::BracketClose),
        Token::InterpolationStart => wrapped_interpolated(stream, TokenDiscriminants::BraceClose),
        Token::SquareBracketOpen => Ok(Expression::Array(ArrayExpression::parse(stream)?)),
        Token::BraceOpen => Ok(Expression::Attrset(AttrsetExpression::parse(stream)?)),
        Token::Rec => Ok(Expression::Attrset(AttrsetExpression {
            recursive: true,
            ..AttrsetExpression::parse(stream)?
        })),
        Token::Dollar => todo!("dollar"),
        Token::String(_string_tokens) => todo!("str"),
        Token::MultilineString(_string_tokens) => todo!("multi"),
        Token::Throw => todo!("throw"),
        Token::Path(_) => todo!("path"),
        Token::BlockComment | Token::InlineComment => Ok(Expression::Comment),
        v => Err(Error::UnexpectedTopLevelToken(v.into())),
    }
}

fn wrapped_interpolated<'a>(
    stream: &mut Peekable<Lexer<'a, Token<'a>>>,
    end: TokenDiscriminants,
) -> Result<Expression<'a>, Error> {
    let node = parse_expression_inner(stream, u8::MAX)?;
    expect_next_token_or_error(stream, end)?;
    Ok(node)
}

#[derive(PartialEq, Debug, Clone)]
pub struct WrappedFloat(f64);

impl Eq for WrappedFloat {}

impl PartialOrd for WrappedFloat {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for WrappedFloat {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.0.total_cmp(&other.0)
    }
}

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub enum BindingName<'a> {
    Lazy(Expression<'a>),
    Value(&'a str),
}

fn parse_binding_name<'a>(
    stream: &mut Peekable<Lexer<'a, Token<'a>>>,
) -> Result<Vec<BindingName<'a>>, Error> {
    let mut name = Vec::new();

    loop {
        match stream.next() {
            Some(Ok(Token::Ident(ident))) => name.push(BindingName::Value(ident)),
            Some(Ok(Token::InterpolationStart)) => {
                name.push(BindingName::Lazy(parse_expression_inner(stream, u8::MAX)?));
                expect_next_token_or_error(stream, TokenDiscriminants::BraceClose)?;
            }
            Some(Ok(Token::Equals)) => return Ok(name),
            Some(Ok(token)) => {
                return Err(Error::UnexpectedToken(
                    token.into(),
                    vec![
                        TokenDiscriminants::Ident,
                        TokenDiscriminants::InterpolationStart,
                        TokenDiscriminants::Equals,
                    ],
                ));
            }
            Some(Err(())) => panic!(),
            None => return Err(Error::UnexpectedEndOfFile),
        }
    }
}

fn expect_next_token_or_error<'a>(
    stream: &mut Peekable<Lexer<'a, Token<'a>>>,
    kind: TokenDiscriminants,
) -> Result<(), Error> {
    match stream.next() {
        Some(Ok(v)) if TokenDiscriminants::from(&v) == kind => Ok(()),
        Some(Ok(v)) => Err(Error::UnexpectedToken(v.into(), vec![kind])),
        Some(Err(())) => panic!(),
        None => Err(Error::UnexpectedEndOfFile),
    }
}

#[cfg(test)]
mod test {
    use std::fs::read_to_string;

    use insta::{assert_debug_snapshot, glob};
    use nix_lexer::Logos as _;

    use crate::parse_expression;

    #[test]
    fn fixtures() {
        glob!("../fixtures", "*.nix", |path| {
            let input = read_to_string(path).expect("failed to read fixture");
            let lex = nix_lexer::Token::lexer(&input);
            match parse_expression(lex) {
                Ok(v) => assert_debug_snapshot!("fixture", v, &input),
                Err(e) => assert_debug_snapshot!("fixture", e, &input),
            }
        });
    }
}
