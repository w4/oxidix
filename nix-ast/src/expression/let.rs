use std::{collections::BTreeMap, iter::Peekable};

use nix_lexer::{SpannedIter, Token, TokenDiscriminants};

use crate::{
    BindingName, Error, Expression, HandleStreamError, SpannedError, expect_next_token_or_error,
    parse_binding_name, parse_expression_inner,
};

use super::attrset::parse_inherits;

#[derive(PartialEq, Eq, Debug, Clone, PartialOrd, Ord)]
pub struct LetExpression<'a> {
    pub bindings: BTreeMap<Vec<BindingName<'a>>, Expression<'a>>,
    pub inner: Expression<'a>,
}

impl<'a> LetExpression<'a> {
    pub fn parse(
        stream: &mut Peekable<SpannedIter<'a, Token<'a>>>,
    ) -> Result<LetExpression<'a>, SpannedError> {
        let mut bindings = BTreeMap::new();

        let inner = loop {
            match stream.peek().span_error()? {
                (Token::In, _) => {
                    stream.next();
                    break parse_expression_inner(stream, u8::MAX)?;
                }
                (Token::Ident(_), _) => {
                    let name = parse_binding_name(stream)?;
                    let value = parse_expression_inner(stream, u8::MAX)?;
                    expect_next_token_or_error(stream, TokenDiscriminants::Semicolon)?;
                    bindings.insert(name, value);
                }
                (Token::Inherit, _) => {
                    parse_inherits(stream, &mut bindings)?;
                }
                (token, span) => {
                    return Err(Error::UnexpectedToken(
                        token.into(),
                        vec![
                            TokenDiscriminants::In,
                            TokenDiscriminants::Ident,
                            TokenDiscriminants::Inherit,
                        ],
                    )
                    .with_span(span));
                }
            }
        };

        Ok(LetExpression { bindings, inner })
    }
}
