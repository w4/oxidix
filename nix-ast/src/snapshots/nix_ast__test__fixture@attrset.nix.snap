---
source: nix-ast/src/lib.rs
expression: "x: {\n  inherit (x) y;\n  inherit x;\n  z = 5;\n}\n"
input_file: nix-ast/fixtures/attrset.nix
---
Lambda(
    LambdaExpression {
        parameters: Named(
            "x",
        ),
        body: Attrset(
            AttrsetExpression {
                recursive: false,
                bindings: {
                    [
                        Value(
                            "x",
                        ),
                    ]: Ident(
                        "x",
                    ),
                    [
                        Value(
                            "y",
                        ),
                    ]: Ident(
                        "x",
                    ),
                    [
                        Value(
                            "z",
                        ),
                    ]: Int(
                        5,
                    ),
                },
            },
        ),
    },
)
